File: wintest/steps/click.py
"""Click step -- click at explicit coordinates.

Click types: click (left), double_click, right_click, middle_click. For
clicks that locate the target by natural-language description (AI), use
the click_element step instead.
"""

from ._base import StepDefinition, FieldDef


def validate(step, step_num):
    if step.click_x is None or step.click_y is None:
        return [f"Step {step_num}: 'click' requires click_x and click_y coordinates"]
    if step.click_type not in ("click", "double_click", "right_click", "middle_click"):
        return [f"Step {step_num}: invalid click_type '{step.click_type}'"]
    return []


def execute(step, agent):
    return agent.click_at(step, click_type=step.click_type or "click")


definition = StepDefinition(
    name="click",
    description="Click at coordinates (left, double, right, or middle click)",
    fields=[
        FieldDef("click_x", "number", required=True),
        FieldDef("click_y", "number", required=True),
        FieldDef("click_type", "string"),
    ],
    validate=validate,
    execute=execute,
)

File: wintest/ui/web/frontend/src/docs/click.ts
import type { StepDoc } from './types';

export const clickDoc: StepDoc = {
  name: 'click',
  title: 'Click',
  summary: 'Click at captured coordinates.',
  description:
    'Performs a click at pixel-exact coordinates (typically recorded by the Test Builder by clicking on the screenshot). Fast, deterministic, no AI required. For clicks that locate a target by natural-language description, use the click_element step.',
  parameters: [
    {
      name: 'click_x',
      type: 'number',
      required: true,
      description: 'Normalized horizontal click coordinate (0.0–1.0).',
    },
    {
      name: 'click_y',
      type: 'number',
      required: true,
      description: 'Normalized vertical click coordinate (0.0–1.0).',
    },
    {
      name: 'click_type',
      type: 'string',
      required: false,
      description:
        'One of "click" (left, default), "double_click", "right_click", or "middle_click".',
    },
  ],
  example:
    '- action: click\n  click_type: double_click\n  click_x: 0.35\n  click_y: 0.62\n  description: "Open the report file"',
};
